blog: add fluent-package v5.2.0 release announcement (#336)
fluent/fluentd-website: content/blog/20241214_fluent-package-v5.2.0-has-been-released.md

# fluent-package v5.2.0 has been released

Hi users!

We have released fluent-package [v5.2.0](https://github.com/fluent/fluent-package-builder/releases/tag/v5.2.0).
fluent-package is a stable distribution package of Fluentd. (successor of td-agent v4)

> This is a maintenance release of v5.x series.

Reviewer: ```suggestion
This release is a new release of v5.2 series.
```

## Changes from fluent-package v5.1.0
* Support upgrade fluentd service with zero-downtime
* Update ruby to 3.2.6
* Update bundled Fluentd to v1.18.0
* Update bundled gems overall
* Solved possibility of DoS through "NoMemoryError"
* deb rpm: Fixed to not execute v4 restart migration process unexpectedly
* msi: set `GEM_HOME`/`GEM_PATH` in fluentd.bat
* Fixed SIGABORT error with fluent-plugin-systemd

## Support upgrade fluentd service with zero-downtime

> Previously, when upgrading the fluent-package, you had to stop the ｆfluentd service,

Reviewer: ```suggestion
Previously, when upgrading the fluent-package, you had to stop the fluentd service,
```

install the new version of the fluent-package, then install any necessary plugins,
and finally restart the fluentd service.

> The starting from fluent-package v5.2.0,

Reviewer: ```suggestion
Starting from fluent-package v5.2.0,
```

we have introduced a new feature that automatically installs the required plugins during the upgrade process,
and introduced the zero-downtime restart feature which allows you to upgrade fluentd without interrupting the logging service.

> The zero-downtime restart feature can be configured by `FLUENT_PACKAGE_SERVICE_RESTART` environemnt variable.

Reviewer: ```suggestion
The zero-downtime restart feature can be configured by `FLUENT_PACKAGE_SERVICE_RESTART` environment variable.
```

Please refer to following section for more details.

* Add `FLUENT_PACKAGE_SERVICE_RESTART` environmental variable
  * This section explains `FLUENT_PACKAGE_SERVICE_RESTART` environment variable.
* Automate Plugin Install for Update on Demand

>   * This section explains the mechanism that install the plug-ins automatically during the upgrade process.

Reviewer: It would be better to align the wording of `plugin`/`plug-in`.

```suggestion
  * This section explains the mechanism that installs the plugins automatically during the upgrade process.
```

### Add `FLUENT_PACKAGE_SERVICE_RESTART` environmental variable
We have introduced `FLUENT_PACKAGE_SERVICE_RESTART` environment variable to configure the zero-downtime restart feature.
The variable exists in the following file:

* RPM: `/etc/sysconfig/fluentd`
* DEB: `/etc/default/fluentd`

Example:

```
FLUENT_PACKAGE_OPTIONS=""
# Control method to upgrade service (auto/manual) restart
FLUENT_PACKAGE_SERVICE_RESTART=auto
```

#### **auto (default)**
The service automatically restarts with zero-downtime restart feature when all of the following conditions are met:

1. The service was active before updating.
2. The installed and upgrading versions are required 5.2.0 or higher (both sides must support this feature).

The plugins will be automatically reinstalled if needed if your environment has an active online connection.
For more details, please refer to "Automate Plugin Install for Update on Demand".

> **Note:** If the service was active before the update, it will not restart automatically. You need to restart the service manually.

Reviewer: > **Note:** If the service was active before the update, it will not restart automatically. You need to restart the service manually.

Is not this an explanation for `manual`?

Author: indeed

1. Collect the current list of plugins before the update.
   - Recognize gems as plugins that has `fluent-plugin-` prefix.
2. After installing the updated package, detect missing plugins.
   - Compare the collected plugin list with the default plugins to identify any missing ones.
3. Install the missing plugins.

#### Manual Plugin Management
If you prefer to manage plugins manually, please set `FLUENT_PACKAGE_SERVICE_RESTART` to `manual`.
This is useful in cases such as:

- Pinning specific plugin versions
- Operating in an offline environment

In this mode, automatic plugin installation and restarts are disabled.
You can manually install plugins and send a `SIGUSR2` signal to trigger the zero-downtime restart feature after the update.

Please refer to [#713](https://github.com/fluent/fluent-package-builder/pull/713) for more details.

## Solved possibility of DoS through "NoMemoryError"
In the previous versions of msgpack gem, there was a possibility that cause DoS (Denial Of Service)
when crafted message was sent to Fluentd.

When that message was sent, it might cause "NoMemoryError" on Fluentd because there is a case that msgpack
try to pre-allocate huge amount of memories.

> Above behavior was fixed not to pre-allocate over 32k entries (limit pre-allocated amount of memories) in msgpack 1.7.3 and bundled into fluent-package v5.0.5.

Reviewer: ```suggestion
Above behavior was fixed not to pre-allocate over 32k entries (limit pre-allocated amount of memories) in msgpack 1.7.3 and bundled.
```
